app/grandchallenge/components/models.py

```diff
@@ -2,19 +2,142 @@
 from typing import Tuple, Type
 
 from django.conf import settings
+from django.contrib.postgres.fields import JSONField
 from django.core.files import File
 from django.db import models
 from django.db.models import Avg, F
+from django.utils._os import safe_join
 from django.utils.text import get_valid_filename
 from django.utils.timezone import now
+from django.utils.translation import gettext_lazy as _
+from django_extensions.db.fields import AutoSlugField
 
+from grandchallenge.cases.models import Image
 from grandchallenge.components.backends.docker import Executor
-from grandchallenge.components.tasks import execute_job
-from grandchallenge.core.storage import private_s3_storage
+from grandchallenge.components.tasks import execute_job, validate_docker_image
+from grandchallenge.components.validators import validate_safe_path
+from grandchallenge.core.storage import (
+    private_s3_storage,
+    protected_s3_storage,
+)
 from grandchallenge.core.validators import ExtensionValidator
 
 
-class ComponentQuerySet(models.QuerySet):
+class InterfaceKindChoices(models.TextChoices):
+    STRING = "STR", _("String")
+    INTEGER = "INT", _("Integer")
+    FLOAT = "FLT", _("Float")
+    BOOL = "BOOL", _("Bool")
+
+    # Annotation Types
+    TWO_D_BOUNDING_BOX = "2DBB", _("2D bounding box")
+    MULTIPLE_TWO_D_BOUNDING_BOXES = "M2DB", _("Multiple 2D bounding boxes")
+    DISTANCE_MEASUREMENT = "DIST", _("Distance measurement")
+    MULTIPLE_DISTANCE_MEASUREMENTS = (
+        "MDIS",
+        _("Multiple distance measurements"),
+    )
+    POINT = "POIN", _("Point")
+    MULTIPLE_POINTS = "MPOI", _("Multiple points")
+    POLYGON = "POLY", _("Polygon")
+    MULTIPLE_POLYGONS = "MPOL", _("Multiple polygons")
+
+    # Choice Types
+    CHOICE = "CHOI", _("Choice")
+    MULTIPLE_CHOICE = "MCHO", _("Multiple choice")
+
+    # Image types
+    IMAGE = "IMG", _("Image")
+    SEGMENTATION = "SEG", _("Segmentation")
+    HEAT_MAP = "HMAP", _("Heat Map")
+
+    # Legacy support
+    JSON = "JSON", _("JSON file")
+    CSV = "CSV", _("CSV file")
+    ZIP = "ZIP", _("ZIP file")
+
+
+class ComponentInterface(models.Model):
+    Kind = InterfaceKindChoices
+
+    title = models.CharField(
+        max_length=255,
+        help_text="Human readable name of this input/output field.",
+        unique=True,
+    )
+    slug = AutoSlugField(populate_from="title")
+    description = models.TextField(
+        blank=True, help_text="Description of this input/output field.",
+    )
+    default_value = JSONField(
+        null=True,
+        default=None,
+        help_text="Default value for this field, only valid for inputs.",
+    )
+    kind = models.CharField(
+        blank=False,
+        max_length=4,
+        choices=Kind.choices,
+        help_text="What kind of field is this interface?",
```

Comment:
I find the word interface in the help text a bit confusing, isn't it the entity's kind?

Reply:
Agreed, can you think of a better way to phrase this? I wanted to use `type` rather than `kind`, but this is a reserved word in python.

Reply:
I guess for the help text we could use type right? so maybe "What kind (or type) of field does this interface describe?"